### ISSUE TITLE
HTTP client tutorial

### ISSUE THREAD

**Issue by snoyberg** [commercialhaskell/jump#6]
Definitely a common case that we should cover. Something to consider while writing: do we want to augment the http-client API with more high-level combinators, or possibly an alternative to record syntax (via setters/getters or lenses) for modifying requests and accessing responses?

I think @nh2 and @bitonic may have had some thoughts on possible API improvements.

**Comment by dpwiz** [comment]
> an alternative to record syntax (via setters/getters or lenses)

Isn't that a niche for `wreq`?

**Comment by bitonic** [comment]
My opinion in this is that having a very easy way to get started with `http-client` is better than just pointing at `wreq` because once you're comfortable with `http-client` you can easily take advantage of the more advanced features (e.g. streaming).

I think an `.Easy` module or something like that might be beneficial, although to be honest I never had that much trouble with `http-client` as it is, so I haven't put much thought into it.

**Comment by snoyberg** [comment]
I agree with @bitonic's comments here. Also, there are some issues with using wreq in non-standard ways (e.g., I saw some bug reports/questions recently around 0-length request bodies). Also, the defaults for wreq are very non-performant (creating a new manager for each request), and the Jump approach should be: recommend things which are easy to use with the right defaults.

One of the reasons I put "improved API" in a documentation comment instead of just an http-client issue is that, based on experience, you can more effectively design a better API when you're trying to teach it to a new user. I'm hoping that, when documenting http-client here, some easy improvements to the API could be discovered that improve usability.

**Comment by dpwiz** [comment]
> Also, the defaults for wreq are very non-performant (creating a new manager for each request), and the Jump approach should be: recommend things which are easy to use with the right defaults.

That makes sense. Wreq looks like a power tool for a user who knows their deal, not a "go fetch me that URL".

**Comment by bitemyapp** [comment]
I used wreq in learning materials because it eliminated a lot of noise. Would it potentially be worth it to have a backup manager singleton that gets torn down on a timer so that this use pattern doesn't bite as much?

The http-client tutorial/examples look good, but the code itself isn't going to convince someone learning Haskell is a good use of their time if they don't otherwise know better. I think most of the time performance doesn't even really matter for what people'll use an http-client for - most aren't in fintech and a lot of people are really just learning and tinkering for the first time.

We need to be able to at least match Python on basic stuff like this:

![screenshot from 2016-04-13 15-02-10](https://cloud.githubusercontent.com/assets/320177/14507554/b5dd5e92-0188-11e6-83d8-846a9fa6cccc.png)

**Comment by snoyberg** [comment]
The Manager issue is just one symptom. Overall wreq hides away too much of
the API by default. I've had multiple unfortunate experiences code
reviewing wreq based code. But beyond that, I have strong reservations
about recommending it as I don't have high confidence that it will be
maintained reliably going forward.

Is there anything inherent to wreq that couldn't be merged back upstream?
Is there any advantage gained by not doing so?

**Comment by rvion** [comment]
> Is there anything inherent to wreq that couldn't be merged back upstream?

my 2 cents:
- I think that lens based / monoid based apis tends to be easier to grasp and read.
- I think it would be practical to have less imports to write down 
- I think some function names could be made more explicit in http-client:  buildRequestFromUrl instead of parseUrl, etc (buildRequestFromUrl is not the best example, because it's long and ugly, but I think parseUrl is hard to follow)
- I think some more integration to aeson would be better  / multipart would be great. (dedicated functions, builders)

> Overall wreq hides away too much of the API by default.

I think there is some better compromise to find: if we had those 3 layers :
    - building requests
    - building managers
    - using managers to produce responses from requests and consume them in various ways (strict / streaming / etc.).

it would'nt hide much of underlying api, yet it would be simpler to follow / use.

I thinka better mix of low level / high level functions is possible.

---

I think the Turtle library is a good example of batteries included library that is not a massive prelude reexporting non-conflicting things, but rather a carefully designed set of modules trying to be coherent. It exposes a nice set of high level functions, but often along notes and more low level functions so implementation is not too far away

https://hackage.haskell.org/package/turtle-1.2.5/docs/Turtle-Prelude.html#g:3
there is nothing about network here, but the spirit is nice I think

(edited)

**Comment by bitemyapp** [comment]
@snoyberg I think lens can be dropped and you could still have a nice API, but the JSON support is a good idea. People aren't often pulling something down that isn't JSON now.

I like other things @rvion is saying, but I'd make the point that job number one for getting people rolling is, "are simple things simple?" They're going to judge a _lot_ on that and there's no reason we can't hit that out of the park, but we have to commit to making that a priority. This can be done without compromising expert modes of use.

I get that you've had bad experiences with poor performing code. That's worth talking about,. Rather than making that decision for users, why not mention the caveats alongside the simple syntax? Explain it could be slower and then show them a benchmark with data so they can decide for themselves whether the overhead is worth it or not.

It's also worth considering whether a mitigation like the singleton I mentioned is in order. Cf. http://www.tedunangst.com/flak/post/firefox-vs-rthreads

**Comment by snoyberg** [comment]
I wasn't actually complaining about the performance with wreq, I'm
complaining about incorrect code being written.

As usual, I'm going to ask for specifics. Which part of the tutorial are
too verbose? Let's fix those.

**Comment by snoyberg** [comment]
And yes, creating a global manager seems reasonable and is easy to do.

**Comment by snoyberg** [comment]
Here's my reimagining of the API:

https://github.com/snoyberg/http-client/pull/188/files

I haven't added the accessors/setters yet (though I've enumerated the functionality I want covered). I'm no longer convinced that a lens-based API for that will be superior.

**Comment by rvion** [comment]
@snoyberg 

I find the new [Network.HTTP.Simple](https://github.com/snoyberg/http-client/blob/better-api/http-conduit/Network/HTTP/Simple.hs) really good !

>  I'm no longer convinced that a lens-based API for that will be superior.

neither do I. The "http client" api here is quite flat (no deeply nested structures), so standard accessor and setters will do just fine. Using (&) notation in the tutorial could even help clean things a bit and show how to _"compose"_ setters.

wreq used notations like :

``` haskell
opts = defaults
  & param "q" .~ ["tetris"]
  & param "language" .~ ["haskell"]
```

but I think

``` haskell
opts = defaults
  & setQueryParam "q" ["tetris"]
  & setQueryParam "language" ["haskell"]
```

is better. function names are clear, etc.

this being said, the wreq tutorial also show json consuming examples:

``` haskell
r ^.. responseBody . key "items" . values .
            key "owner" . key "login" . _String
```

and I think that `Network.HTTP.Simple` tutorial could also show one exemple like this. Indeed, lots of people trying to use Network.HTTP.Simple will also need to play with complex json. that's a good way to provide a handy examples and to link the concepts / librairies and documentation pages, 

some more remarks:
- Network.HTTP.Simple doesn't contain the "Client" word in it, should it ?

---

### about naming

I just want to share some experience with naming and preludes.

some time ago, I build a "qualified module reexporter" called [jetpack-gen](https://github.com/rvion/ride/tree/master/jetpack-gen) (never released, never announced) that was taking as input lists of modules ([directly from the readme](https://github.com/rvion/ride/blob/655003916da9629b96e537ae809053e6d2e406b6/jetpack/README.md)):

```
m Data.Map.Strict
c Data.Conduit
cl Data.Conduit.List
cb Data.Conduit.Binary
```

and was outputing a [**giant** package](https://github.com/rvion/ride/tree/655003916da9629b96e537ae809053e6d2e406b6/jetpack) reexporting everything:
![image](https://cloud.githubusercontent.com/assets/2150990/14519037/9315fd12-021c-11e6-89e7-d85ce7a6097e.png)

if you look at my own [Network.HTTP.Client.AsHttp](https://github.com/rvion/ride/blob/655003916da9629b96e537ae809053e6d2e406b6/jetpack/src/Network/HTTP/Client/AsHttp.hs#L141) here is it :)

``` haskell
-- [...]
-- http_applyBasicProxyAuth :: ByteString -> ByteString -> Request -> Request
http_applyBasicProxyAuth = I.applyBasicProxyAuth

-- http_getUri :: Request -> URI
http_getUri = I.getUri

-- http_observedStreamFile :: (StreamFileStatus -> IO ()) -> FilePath -> IO RequestBody
http_observedStreamFile = I.observedStreamFile

-- http_parseUrl :: forall (m :: * -> *). MonadThrow m => String -> m Request
http_parseUrl = I.parseUrl

-- http_setQueryString :: [(ByteString, Maybe ByteString)] -> Request -> Request
http_setQueryString = I.setQueryString

-- http_streamFile :: FilePath -> IO RequestBody
http_streamFile = I.streamFile

-- http_urlEncodedBody :: [(ByteString, ByteString)] -> Request -> Request
http_urlEncodedBody = I.urlEncodedBody

-- [...]

type HttpManagerSettings  = I.ManagerSettings
get_http_managerConnCount o = I.managerConnCount o
set_http_managerConnCount x o = o { I.managerConnCount = x}
get_http_managerRawConnection o = I.managerRawConnection o
set_http_managerRawConnection x o = o { I.managerRawConnection = x}
-- [...]


```

the whole point was to generate a non clashing set of thouthands of usefull functions from one same module: here is the [full symbol list](https://github.com/rvion/ride/blob/master/jetpack/full-exported-symbol-list.txt#L3673), here is the reexporting [top level module](https://github.com/rvion/ride/blob/655003916da9629b96e537ae809053e6d2e406b6/jetpack/src/Exports.hs)
I used it for some experiment and it was IMO really pleasant.
it was the first time haskell was not annoying me with imports.
This was really alpha software, code quality is not the best, and there was still a lot of small problems.. but: even for some alpha stuff, I was really productive with it. Sadly, I never took the time to write down all my notes and plans about it.

**Comment by snoyberg** [comment]
Just to test the waters a bit, I pushed some commits with lens-based access to the `Request`. Here's my take-away:
- I hadn't realized that it's so awkward to have write-only fields (no reading) with lens, somehow I've missed that in my usage until now. That's definitely an annoyance for things like turning off status code checking
- Avoiding API duplication with `getRequestMethod` and `setRequestMethod`, and instead just using `requestMethod` for both _is_ nice, as expected. How nice? Not sure
- Providing a tutorial on "lens + aeson" should IMO be a separate tutorial here, with a prominent link at the end of the http-client tutorial. Such functionality is not specific to making HTTP requests (e.g., it's useful when parsing config files)

**Comment by rvion** [comment]
I can't seem to find the commits, did you push them on github ?

**Comment by snoyberg** [comment]
Pushed now, sorry

**Comment by rvion** [comment]
> I hadn't realized that it's so awkward to have write-only fields (no reading) with lens, somehow I've missed that in my usage until now. That's definitely an annoyance for things like turning off status code checking
> 
> Avoiding API duplication with getRequestMethod and setRequestMethod, and instead just using requestMethod for both is nice, as expected. How nice? Not sure

Not sure either.

I you use lens:

why not

```
type Setter' s a = forall f. Functor f => (a -> Indentity a) -> s -> Indentity s
```

instead of Lens' for setter ? (
- lens: http://hackage.haskell.org/package/lens-4.13.2.1/docs/Control-Lens-Setter.html#t:Setter-39-
- microlens: http://hackage.haskell.org/package/microlens-0.4.3.0/docs/Lens-Micro.html#t:ASetter
  )
- you could then remove the 6 `-- This lens does not allow inspecting the request body`
- it would partially fix the problem with hard to read `requestBasicAuth :: Lens H.Request H.Request () (S.ByteString, S.ByteString)`

on the other side, without lenses
- you can provide smart setters (taking 2 parameters for instance) instead of having to bundle those in tupples with lens
- it's still easy to use the `to` lens  (`to getRequestQueryString`) to chain simple getters and setters in lens based code

so either seems fine by me.
I used to want lens everywhere, but simplicity seems more important now

other remark: one could imagine having both: get/set versions + prefix-less lens.
I'm not saying this is a good idea, but it is possible

> Providing a tutorial on "lens + aeson" should IMO be a separate tutorial here, with a prominent link at the end of the http-client tutorial. Such functionality is not specific to making HTTP requests (e.g., it's useful when parsing config files)

"a prominent link" would sure be enough :)

(edited)

**Comment by snoyberg** [comment]
If you can figure out how to make an `ASetter` or similar for those lenses, that would be great. I couldn't get it to work.

**Comment by snoyberg** [comment]
Latest status: the branch is updated to not use lenses at all, but instead direct getters and setters. Very little of value was lost (besides some shorter names and less API-level noise around duplicated functions), we don't have any "maybe a lens" concerns, and I think this is much more beginner friendly. So unless someone comes up with a brilliant lens-based answer, I think the lens experiment is officially over.

**Comment by bitemyapp** [comment]
I think you're right that it's better to just use direct getters and setters in this case. I don't think nested updating or prisms were super-duper needed here. People can use lens-aeson and the like on top of this without much fuss and that can be an example in its own right elsewhere.

**Comment by snoyberg** [comment]
If really desired, I could see providing one lonely responseBody lens for
really easy composition with lens-aeson.

**Comment by rvion** [comment]
I also think it's all fine like this.
I would even -1 a responseBody lens. One can still wrap getResponseBody / setResponseBody easily to provide lens compatibility. (or chain stuff with `xxx . to getResponseBody . yyy`)

**Comment by bitemyapp** [comment]
I use `to` a lot in my code, works fine for this purpose.

**Comment by snoyberg** [comment]
Alright, I think this is ready to go. I've opened PR #10 to merge in the updated docs. I'll do that once the new API is available in an LTS Haskell release.
